### PR TITLE
Set up eth1 monitor component

### PR DESCRIPTION
**Pull request by NIC619**
### How was it fixed?
Set up eth1 monitor so that beacon node can request `Eth1Data` and `Deposit` from eth1 monitor.
This is the first part: set up eth1 monitor component.
- [x] Abstract the eth1 data provider.
    - For local running testnet, a fake eth1 data provider will be used.
        - It deterministically generates eth1 block(block_hash, block_number and timestamp)
    - For interop testnet, provide the eth1 client RPC endpoint info via `--eth1client_rpc` when running `trinity-beacon`
- [x] Add `GetDistanceRequest`/`Response` to `Eth1Monitor`

**NOTE: There are some dirty hacks to get the fake eth1 data provider running. Will get rid of them in subsequent PRs, e.g.,**
- hardcoded initial deposits
    - read validator keys from `eth2/beacon/scripts/quickstart_state/keygen_16_validators.yaml` and generate those initial deposits
        - eth1 monitor need these deposit data but it's not provided in the genesis state
- hardcoded parameters like `START_BLOCK_NUMBER`, `START_BLOCK_TIMESTAMP`
- currently `NUM_DEPOSITS_PER_BLOCK` is set to `0` so no new deposits will be made

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

### CODE REVIEW

File: trinity/components/eth2/eth1_monitor/eth1_data_provider.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    @abstractmethod
+    def get_deposit_count(self, block_number: BlockNumber) -> bytes:
+        ...
+
+    @abstractmethod
+    def get_deposit_root(self, block_number: BlockNumber) -> Hash32:
+        ...
+
+
+AVERAGE_BLOCK_TIME = 20
```
Comment (hwwhww) [inline trinity/components/eth2/eth1_monitor/eth1_data_provider.py]: Maybe add a comment to note that it's only for `FakeEth1DataProvider`.

File: trinity/components/eth2/eth1_monitor/eth1_data_provider.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+
+class FakeEth1DataProvider(BaseEth1DataProvider):
```
Comment (hwwhww) [inline trinity/components/eth2/eth1_monitor/eth1_data_provider.py]: Maybe show `Web3Eth1DataProvider` before `FakeEth1DataProvider`?

File: trinity/components/eth2/eth1_monitor/component.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        #     w3: Web3 = None
+
+        # TODO: For now we use fake eth1 monitor. So we load validators data from
+        # interop setting and hardcode the deposit data into fake eth1 data provider.
+        chain = chain_config.beacon_chain_class(
+            base_db,
+            chain_config.genesis_config,
+        )
+        config = chain.get_state_machine().config
+        # Below code snippet is copied from generate_beacon_genesis.py
```
Comment (hwwhww) [inline trinity/components/eth2/eth1_monitor/component.py]: I'd say
(1) Refactor `generate_beacon_genesis.py`, put the shared logic to `trinity/eth2/beacon/tools/builder/initializer.py`.
(2) Import function from `trinity/eth2/beacon/tools/builder/initializer.py`.

File: trinity/components/eth2/eth1_monitor/eth1_data_provider.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    ) -> None:
+        self.start_block_number = start_block_number
+        self.start_block_timestamp = start_block_timestamp
+        self.num_deposits_per_block = num_deposits_per_block
+        self.initial_deposits = initial_deposits
+        self.num_initial_deposits = len(initial_deposits)
+
+    # TODO: Remove this if we no longer need a fake provider
+    @property
+    def is_fake_provider(self) -> bool:
```
Comment (mhchia) [inline trinity/components/eth2/eth1_monitor/eth1_data_provider.py]: As we discussed offline before, it seems better to remove `is_fake_provider` and make `FakeEth1DataProvider` keep the whole deposit data.

File: trinity/components/eth2/eth1_monitor/eth1_data_provider.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+            self._deposit_contract.events.DepositEvent._get_event_abi()
+        )
+        self._deposit_event_topic = encode_hex(
+            event_abi_to_log_topic(self._deposit_event_abi)
+        )
+
+    def get_block(self, arg: Union[Hash32, int, str]) -> Optional[Eth1Block]:
+        block_dict = self.w3.eth.getBlock(arg)
+        if block_dict is None:
+            return None
```
Comment (ChihChengLiang) [inline trinity/components/eth2/eth1_monitor/eth1_data_provider.py]: Should we do the raising exception style in this API?

File: trinity/components/eth2/eth1_monitor/eth1_monitor.py [hunk 6]
```diff
@@ -281,15 +277,22 @@ def _get_deposit(self, deposit_count: int, deposit_index: int) -> Deposit:
         `latest_block.number - self._num_blocks_confirmed`.
         """
         while True:
-            block = _w3_get_block(self._w3, "latest")
+            block = self._eth1_data_provider.get_block("latest")
+            if block is None:
+                raise Eth1MonitorValidationError(f"Fail to get latest block")
```
Comment (ChihChengLiang) [inline trinity/components/eth2/eth1_monitor/eth1_monitor.py]: Unused f string

**Pull request merged**